BitVector: added support for reverse iteration

tquatmann opened the pull request:
Also use builtin functions to find correct bit indices more efficiently.

volkm reviewed (approved):
I have some smaller comments. Overall it looks fine to me.

tquatmann commented on src/storm/storage/BitVector.h:
```diff
@@ -532,6 +632,28 @@ class BitVector {
+     * bit at this index itself is *not* included in the search range.
+     * @return The smallest index i such that all bits in the range [i,endIndex) are 0.
+     */
+    uint64_t getStartOfZeroSequenceBefore(uint64_t endIndex) const;
```
I also had this in mind. The problem is that we return the index **after** the `1` . 
With the current name, it is more clear what happens when the vector consists of only `0`: we return the start of the `0` sequence, i.e. index 0.
With `getBackwardsNextSetIndex` I would expect that the result always points to a `1`. However, when there is no `1` left of `endIndex` it is very unclear what should be returned.

volkm replied:
Makes sense. Thanks for the clarification.

tquatmann commented on src/storm/storage/BitVector.cpp:
```diff
@@ -704,90 +753,128 @@ BitVector::const_iterator BitVector::begin() const {
+            if constexpr (Backward) {
+                if (currentBucketIndexOffset <= startingIndex) {
+                    // No bucket found!
+                    return startingIndex;
```
Turns out the doxygen was outdated, good find! 
In backwards mode, `startIndex` serves as the "bit not found" result which is valid because we always return the found index + 1.  I also fixed the `startingIndex >= endIndex` case at the beginning.

volkm replied:
Great, thanks!

tquatmann commented on src/storm/storage/BitVector.cpp:
```diff
@@ -704,90 +753,128 @@ BitVector::const_iterator BitVector::begin() const {
+#if (defined(__GNUG__) || defined(__clang__))
+    // Use fast and easy builtin functions to find the correct bit index
+    if (Backward) {
+        return std::max<uint64_t>(startingIndex,
```
The startIndex might point somewhere into the current bucket so the found bit might come before the startIndex. In this case we want to return `startIndex`

volkm replied:
Of course, makes sense.

tquatmann commented:
Thanks for the helpful comments :)
Sorry, I replaced my commit so its hard to see what changed: I updated the documentation, added some clarifying comments, and added `constexpr`. We also now return startIndex if Backward is true here:

```c++
uint_fast64_t BitVector::getNextIndexWithValue(uint64_t const* dataPtr, uint64_t startingIndex, uint64_t endIndex) {
    if (startingIndex >= endIndex) {
        return Backward ? startingIndex : endIndex;
    }
...
```

tquatmann commented:
FYI: this test just time-outed for CLN exact; GMP rational functions; Spot, Debug:
```
[----------] 6 tests from DtmcPrctlModelCheckerTest/24, where TypeParam = (anonymous namespace)::DdSylvanRationalSearchEnvironment
[ RUN      ] DtmcPrctlModelCheckerTest/24.Die
[       OK ] DtmcPrctlModelCheckerTest/24.Die (369 ms)
[ RUN      ] DtmcPrctlModelCheckerTest/24.Crowds
```
I assume/hope that this is unrelated to the changes made here.

volkm commented on src/storm/storage/BitVector.h:
```diff
@@ -564,16 +686,20 @@ class BitVector {
+     * @param endIndex The smallest index that is not part of the search range. Should not be larger than the total size of the BitVector.
+     * @return If backwards is false, the index of the found bit is returned (or endIndex in case no bit with Value is present).
+     *         If backwards is true, the index **after** the found bit is returned (or startIndex in case no bit with value Value is present).
+     *         If startIndex >= endIndex, we always return endIndex.
```
Sorry to be so nitpicky, but can you update the documentation? It returns endIndex for forwards and startIndex for backwards.

tquatmann replied:
Thanks 🥲

volkm reviewed (approved):
I have one last minor thing regarding the documentation. Otherwise you can merge it. Thanks for your changes!
